### DISCUSSION TITLE
Important concept: Are llama_context and llama_batch shared by multiple users?

### DISCUSSION THREAD

**Discussion by jygmysoul** [ggml-org/llama.cpp#8175] [Q&A]
Important concept: Are llama_context and llama_batch shared by multiple users?
In the server scenario, when providing reasoning for multiple clients, is a llama_context and llama_batch created for each user?
Or are multiple clients sharing one llama_context and llama_batch? I saw why the server.cpp example shared one. This is very critical.
I came here to ask the official to verify:

struct server_context {
llama_model * model = nullptr;
llama_context * ctx = nullptr;
gpt_params params;
llama_batch batch;
.......
}

I understand that there should be one llama_context and llama_batch for each client. This logic is very clear.
Am I wrong? Please give an accurate answer from the official staff. Thank you. It is very critical.

**Comment by ggerganov** [comment]
The design is to use one `llama_context` + `llama_batch` to handle multiple clients. See the [parallel](https://github.com/ggerganov/llama.cpp/tree/master/examples/parallel) example to understand how this works
